Sources/NIOCore/IPAddresses.swift

```diff
@@ -0,0 +1,577 @@
+public struct IPv4Bytes {
+    public var bytes: IPv4BytesTuple
+    
+    public init(_ bytes: IPv4BytesTuple) {
+        self.bytes = bytes
+    }
+}
+
+/// Represent the bytes for an `IPv6Address.
+public struct IPv6Bytes {
+    public var bytes: IPv6BytesTuple
+    
+    public init(_ bytes: IPv6BytesTuple) {
+        self.bytes = bytes
+    }
+}
+
+/// Represent a single `IPAddress`.
+public enum IPAddress: CustomStringConvertible {
+    /// A single IPv4 address for `IPAddress`.
+    public struct IPv4Address: CustomStringConvertible {
+        /// The bytes storing the address of the IPv4 address.
+        public var address: IPv4Bytes
+        
+        /// Get the `IPv4Address` as a string.
+        public var ipAddressString: String {
+            self.address.lazy.map({String($0)}).joined(separator: ".")
+        }
+        
+        /// A human-readable description of this `IPv4Address`. Mostly useful for logging.
+        public var description: String {
+            return "[IPv4]\(self.ipAddressString)"
+        }
+        
+        /// Get the libc address for an IPv4 address.
+        public var posix: in_addr {
+            get {
+                return in_addr.init(s_addr: UInt32(uint8Tuple: self.address.bytes))
+            }
+        }
+        
+        /// Creates a new `IPv4Address`.
+        ///
+        /// - parameters:
+        ///   - address: Bytes that hold the IPv4 address.
+        public init(address: IPv4Bytes) {
+            self.address = address
+        }
+        
+        /// Creates a new `IPv4Address`.
+        ///
+        /// - parameters:
+        ///   - packedBytes: Collection of UInt8 that holds the address.
+        public init<Bytes: Collection>(packedBytes bytes: Bytes) where Bytes.Element == UInt8 {
+            var ipv4Bytes = IPv4Bytes((0,0,0,0))
+            
+            for (idx, elt) in bytes.prefix(4).enumerated() {
+                ipv4Bytes[idx] = elt
+            }
+            
+            self = .init(address: ipv4Bytes)
+        }
+        
+        /// Creates a new `IPv4Address`.
+        ///
+        /// - parameters:
+        ///   - string: String representation of an IPv4 address.
+        public init(string: String) throws {
+            var bytes: [UInt8] = [0,0,0,0]
+            var idx: Int = 0
+            
+            for char in string.utf8 {
+                if char == UInt8(ascii: ".") {
+                    idx += 1
+                } else if UInt8(ascii: "0")...UInt8(ascii: "9") ~= char {
+                    let number = char - UInt8(ascii: "0")
+                    if idx > 3 || bytes[idx] > 25 || (255 - number < (bytes[idx] * 10)) {
+                        throw IPAddressError.failedToParseIPString(string)
+                    }
+                    bytes[idx] = bytes[idx] * 10 + number
+                } else {
+                    throw IPAddressError.failedToParseIPString(string)
+                }
+            }
+            if idx != 3 {
```

Comment:
I don't think this algorithm is quite right: it would allow `...` as a valid IPv4 address, which it should not. Can we add some test cases that hit these edge cases and confirm that we're doing the appropriate thing?